main.cpp

```diff
@@ -0,0 +1,42 @@
+#include "ip_filter.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <string>
+#include <vector>
+
+template<typename C>
+void apply_filters(C&& c){};
+
+template<typename C, typename Arg, typename ...Args>
+void apply_filters(C&& c, Arg&& arg, Args&&... args) {
+
+  for(auto const& ip: ip_filter::filter{std::forward<Arg>(arg)}(std::forward<C>(c)))
+    std::cout << ip << '\n';
+
+  apply_filters(std::forward<C>(c), std::forward<Args>(args)...);
+}
+
+int main(int, char const *[]) try
+{
+  std::vector<ip_filter::ip> ip_pool;
+
+  for(std::string line; std::getline(std::cin, line); )
+    ip_pool.emplace_back(line.substr(0, line.find('\t')));
+
+  std::sort(std::begin(ip_pool), std::end(ip_pool), std::greater<ip_filter::ip>());
+
+  apply_filters(ip_pool,
+      [](ip_filter::ip const &ip){return true;},
+      [](ip_filter::ip const &ip){return ip[0] == 1;},
+      [](ip_filter::ip const &ip){return ip[0] == 46 && ip[1] == 70;},
+      [](ip_filter::ip const &ip){return ip[0] == 46 || ip[1] == 46 ||
+                                         ip[2] == 46 || ip[3] == 46;}
```

Comment:
Cute smart! But  don't you think it's far distant from:
filter(1)
filter(46,70)
filter_any(46) <--- especially for this!!!! Checking every numer in main func loks odd.
Why don't you write some wrapper?

Reply:
>Checking every numer in main func loks odd.

I will add iterator support and then will use std::any_of in this case

>Why don't you write some wrapper?

Well,  maybe it's a good idea in some way. But can you understand what it means
`filter(1)` or `filter(46, 70) from code? As for me, this weird numbers tells nothing. 
That's why lambdas w/o wrappers are better in this case.